action/protocol/staking/vote_reviser.go

```diff
@@ -14,39 +14,74 @@ import (
 
 // VoteReviser is used to recalculate candidate votes.
 type VoteReviser struct {
-	reviseHeights []uint64
-	cache         map[uint64]CandidateList
-	c             genesis.VoteWeightCalConsts
+	reviseHeights      []uint64
+	cache              map[uint64][2]CandidateList
+	c                  genesis.VoteWeightCalConsts
+	correctCandsHeight uint64
 }
 
 // NewVoteReviser creates a VoteReviser.
-func NewVoteReviser(c genesis.VoteWeightCalConsts, reviseHeights ...uint64) *VoteReviser {
+func NewVoteReviser(c genesis.VoteWeightCalConsts, correctCandsHeight uint64, reviseHeights ...uint64) *VoteReviser {
 	return &VoteReviser{
-		reviseHeights: reviseHeights,
-		cache:         make(map[uint64]CandidateList),
-		c:             c,
+		reviseHeights:      reviseHeights,
+		cache:              make(map[uint64][2]CandidateList),
+		c:                  c,
+		correctCandsHeight: correctCandsHeight,
 	}
 }
 
 // Revise recalculate candidate votes on preset revising height.
 func (vr *VoteReviser) Revise(csm CandidateStateManager, height uint64) error {
 	if !vr.isCacheExist(height) {
+		var dirty CandidateList
 		cands, err := vr.calculateVoteWeight(csm)
 		if err != nil {
 			return err
 		}
-		vr.storeToCache(height, cands)
+		if vr.needCorrectCands(height) {
+			for _, c := range csm.DirtyView().candCenter.base.nameMap {
+				dirty = append(dirty, c)
+			}
+			for _, c := range csm.DirtyView().candCenter.base.operatorMap {
+				dirty = append(dirty, c)
+			}
+			cands, err = vr.correctAliasCands(csm, cands)
+			if err != nil {
+				return err
+			}
+		}
+		vr.storeToCache(height, dirty, cands)
 	}
 	return vr.flush(height, csm)
 }
 
+func (vr *VoteReviser) correctAliasCands(csm CandidateStateManager, cands CandidateList) (CandidateList, error) {
+	ownerMap := map[string]*Candidate{}
+	for addr, owner := range csm.DirtyView().candCenter.base.ownerMap {
+		ownerMap[addr] = owner
+	}
+	var retval CandidateList
+	for _, c := range cands {
+		if owner, ok := ownerMap[c.Owner.String()]; ok {
+			c.Operator = owner.Operator
+			c.Reward = owner.Reward
+			c.Name = owner.Name
+		}
+		retval = append(retval, c)
+	}
+	return retval, nil
+}
+
 func (vr *VoteReviser) result(height uint64) (CandidateList, bool) {
 	cands, ok := vr.cache[height]
-	return cands, ok
+	if !ok {
+		return nil, false
+	}
+	return cands[1], true
 }
 
-func (vr *VoteReviser) storeToCache(height uint64, cands CandidateList) {
-	vr.cache[height] = cands
+func (vr *VoteReviser) storeToCache(height uint64, dirty, cands CandidateList) {
+	vr.cache[height] = [2]CandidateList{dirty, cands}
```

Comment:
cache should only contain the correct candidates, we should not need to change the implementation here